File: tree/find_mode.go
package tree

func findMode(root *TreeNode) []int {
	var res []int
	max, cnt := 0, 1
	var pre *TreeNode
	inOrder(root, &pre, &cnt, &max, &res)
	return res
}

func inOrder(node *TreeNode, pre **TreeNode, cnt *int, max *int, res *[]int) {
	if node == nil {
		return
	}
	inOrder(node.Left, pre, cnt, max, res)
	q := &pre
	n := **q
	if n != nil && n.Val == node.Val {
		*cnt++
	} else {
		*cnt = 1
	}
	output := *res
	if *cnt >= *max {
		if *cnt > *max {
			output = []int{}
		}
		*res = append(output, node.Val)
		*max = *cnt
	}
	**q = node
	inOrder(node.Right, pre, cnt, max, res)
}

File: src/tree/find_mode.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
import {TreeNode} from "./tree_node";
export function findMode(root: TreeNode | null): number[] {
    let o:{
        max: number
        cnt: number
    } = {
        max:0,
        cnt:1
    };
    let pre:TreeNode[] = [];
    let res:number[] = [];
    inOrder(root, pre,o,res);
    return res;
}

function inOrder(node: TreeNode|null, pre: TreeNode[], o:{ max: number; cnt: number }, res: number[]): number[]|undefined{
    if (node==null) {
        return;
    }
    inOrder(node.left, pre, o, res);
    let preItem = pre.pop()

    if (preItem!=undefined) {
        o.cnt = (node.val == preItem.val) ? o.cnt + 1 : 1;
    }
    if (o.cnt >= o.max) {
        if (o.cnt>o.max) {
            res.length = 0
        }
        res.push(node.val)
        o.max = o.cnt
    }
    pre.push(node)
    inOrder(node.right, pre, o, res);
}